packages/react-scripts/bin/react-scripts.js

```diff
@@ -8,6 +8,163 @@
 
+// Makes the script crash on unhandled rejections instead of silently
+// ignoring them. In the future, promise rejections that are not handled will
+// terminate the Node.js process with a non-zero exit code.
+process.on('unhandledRejection', err => {
+  throw err;
+});
+
+// Set this so that importing "../config/env" doesn't fail.
+// In practice it won't matter because we'll spawn another process now.
+process.env.NODE_ENV = 'development';
+// Ensure environment variables are read.
+require('../config/env');
+
+const chalk = require('chalk');
+const fs = require('fs');
+const path = require('path');
+
+// We assume that having wrong versions of these
+// in the tree will likely break your setup.
+// This is a relatively low-effort way to find common issues.
+function verifyPackageTree() {
+  const depsToCheck = [
+    // These are packages most likely to break in practice.
+    // See https://github.com/facebookincubator/create-react-app/issues/1795 for reasons why.
+    // I have not included Babel here because plugins typically don't import Babel (so it's not affected).
+    'eslint',
+    'jest',
+    'webpack',
+    'webpack-dev-server',
+  ];
+  // Inlined from semver-regex, MIT license.
+  // Don't want to make this a dependency after ejecting.
+  const getSemverRegex = () =>
+    /\bv?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-[\da-z-]+(?:\.[\da-z-]+)*)?(?:\+[\da-z-]+(?:\.[\da-z-]+)*)?\b/gi;
+  const ownPackageJson = require('../package.json');
+  const expectedVersionsByDep = {};
+  // Gather wanted deps
+  depsToCheck.forEach(dep => {
+    const expectedVersion = ownPackageJson.dependencies[dep];
+    if (!expectedVersion) {
+      throw new Error('This dependency list is outdated, fix it.');
+    }
+    if (!getSemverRegex().test(expectedVersion)) {
+      throw new Error(
+        `The ${dep} package should be pinned, instead got version ${expectedVersion}.`
+      );
+    }
+    expectedVersionsByDep[dep] = expectedVersion;
+  });
+  // Verify we don't have other versions up the tree
+  let currentDir = __dirname;
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const previousDir = currentDir;
+    currentDir = path.resolve(currentDir, '..');
+    if (currentDir === previousDir) {
+      // We've reached the root.
+      break;
+    }
+    const maybeNodeModules = path.resolve(currentDir, 'node_modules');
+    if (!fs.existsSync(maybeNodeModules)) {
+      continue;
+    }
+    depsToCheck.forEach(dep => {
+      const maybeDep = path.resolve(maybeNodeModules, dep);
+      if (!fs.existsSync(maybeDep)) {
+        return;
+      }
+      const maybeDepPackageJson = path.resolve(maybeDep, 'package.json');
+      if (!fs.existsSync(maybeDepPackageJson)) {
+        return;
+      }
+      const depPackageJson = JSON.parse(
+        fs.readFileSync(maybeDepPackageJson, 'utf8')
+      );
+      const expectedVersion = expectedVersionsByDep[dep];
+      if (depPackageJson.version !== expectedVersion) {
```

Comment:
Assuming the package is found and the version matches, do we need to continue looking up the tree?
In this case, we could `splice` it out of the array.

Our dependencies should never be hoisted out of our project [except in the case of Yarn Workspaces] and this fails spectacularly currently.

---

I'm fine with this if we want to be cautious and support Yarn Workspaces in the future

Reply:
It's a shame we have to look multiple levels, because if we only needed the first occurrence we could be sly and use `require`:
```js
depsToCheck.map(name => {
  try {
    const { version } = require(`${name}/package.json`)
  } catch (e) {
    // ignored
  }
})
```

Reply:
We want to support YW so I think we shouldn’t stop.

What fails?

Reply:
Also just one level doesn’t help. In my example we’ll still find “our own” webpack but the project above might have its copy.

Reply:
> What fails?

`react-scripts` when used with Yarn Workspaces, not this PR (what https://github.com/facebookincubator/create-react-app/pull/3435 is attempting to solve).

Reply:
> Also just one level doesn’t help. In my example we’ll still find “our own” webpack but the project above might have its copy.

You're right. This LGTM.